src/fiber/fiber.ml

```diff
@@ -1,284 +1,249 @@
 open! Stdune
 
-type 'a t = ('a -> unit) -> unit
+type 'a t = ('a -> effect) -> effect
+
+and effect =
+  | Read_ivar : 'a ivar * ('a -> effect) -> effect
+  | Fill_ivar : 'a ivar * 'a * (unit -> effect) -> effect
+  | Suspend : ('a k -> unit) * ('a -> effect) -> effect
+  | Restart : 'a k * 'a * (unit -> effect) -> effect
+  | Get_var : 'a Univ_map.Key.t * ('a option -> effect) -> effect
+  | Set_var : 'a Univ_map.Key.t * 'a * (unit -> effect) -> effect
```

Comment:
Remind me again why we need the thunk around `effect` for `Set_var`?

Reply:
The scheduler will add the new binding `var=value` to the current execution context and execute the thunk in this new execution context.

The thunk will eventually produce the `Unwind` effect to go up one execution context, which will remove the binding.